File: demo/earthball.c
#include <stdlib.h>
#define SDL_MAIN_HANDLED
#include <SDL.h>
#include "monocle.h"

void
instructions(void)
{
    MNCL_DATA *insts;
    insts = mncl_data_resource("instructions");
    if (insts && insts->tag == MNCL_DATA_ARRAY) {
        int i;
        printf("Instructions:\n");
        for (i = 0; i < insts->value.array.size; ++i) {
            MNCL_DATA *val = insts->value.array.data[i];
            printf ("    %s\n", (val && val->tag == MNCL_DATA_STRING) ? val->value.string : "(invalid datum)");
        }
    } else {
        printf("No instructions available\n");
    }
}

void
update(MNCL_OBJECT *obj)
{
    float newx = obj->x + obj->dx;
    float newy = obj->y + obj->dy;
    if (newx < 0) {
        obj->x = 0; obj->dx = -obj->dx;
    }
    if (newy < 0) {
        obj->y = 0; obj->dy = -obj->dy;
    }
    if (newx > 704) {
        obj->x = 704; obj->dx = -obj->dx;
    }
    if (newy > 416) {
        obj->y = 416; obj->dy = -obj->dy;
    }
}

int
main(int argc, char **argv)
{
    MNCL_SFX *sfx;
    MNCL_FONT *monospace;
    int done, i, countdown, music_on, bg, music_volume;

    mncl_init();
    mncl_config_video("Earthball Demo", 768, 480, 0, 0);
    mncl_add_resource_zipfile("earthball-res.zip");

    mncl_load_resmap("earthball.json");
    sfx = mncl_sfx_resource("sfx");
    monospace = mncl_font_resource("monospace");
    mncl_play_music_resource("bgm", 2000);

    for (i = 0; i < 16; ++i) {
        MNCL_OBJECT *obj = mncl_create_object(rand() % (768 - 64), rand() % (480 - 64), "earth");
        obj->dx = (rand() % 5 + 1) * ((rand() % 2) ? 1 : -1);
        obj->dy = (rand() % 5 + 1) * ((rand() % 2) ? 1 : -1);
        obj->f = rand() % 30;
    }

    done = 0;
    countdown = 0;
    music_on = 1;
    bg = 0;
    music_volume = 128;

    instructions();

    mncl_hide_mouse_in_fullscreen(1);

    while (!done) {
        MNCL_EVENT *e = mncl_pop_global_event();
        switch(e->type) {
        case MNCL_EVENT_QUIT:
            done = 1;
            break;
        case MNCL_EVENT_KEYDOWN:
            switch (e->value.key) {
            case SDLK_ESCAPE:
                mncl_fade_out_music(1000);
                countdown = 100;
                break;
            case SDLK_b:
                bg = (bg + 1) % 4;
                mncl_set_clear_color(bg == 1 ? 128 : 0, bg == 2 ? 128 : 0, bg == 3 ? 128 : 0);
                break;
            case SDLK_p:
                music_on = 1 - music_on;
                if (music_on) {
                    mncl_resume_music();
                } else {
                    mncl_pause_music();
                }
                break;
            case SDLK_t:
                printf("Toggled fullscreen to: %s\n", mncl_toggle_fullscreen() ? "on" : "off");
                printf("Fullscreen reported as: %s\n", mncl_is_fullscreen() ? "on" : "off");
                break;
            case SDLK_UP:
                music_volume += 16;
                if (music_volume > 128) {
                    music_volume = 128;
                }
                mncl_music_volume(music_volume);
                printf("Music volume now: %d\n", music_volume);
                break;
            case SDLK_DOWN:
                music_volume -= 16;
                if (music_volume < 0) {
                    music_volume = 0;
                }
                mncl_music_volume(music_volume);
                printf("Music volume now: %d\n", music_volume);
                break;
            case SDLK_SPACE:
                mncl_play_sfx(sfx, 128);
                break;
            case SDLK_LSHIFT:
                mncl_play_sfx(sfx, 64);
                break;
            default:
                break;
            }
            break;
        case MNCL_EVENT_PRERENDER:
            if (e->value.self) {
                update(e->value.self);
            } else if (countdown > 0) {
                if (--countdown == 0) {
                    done = 1;
                }
            }
            break;
        case MNCL_EVENT_RENDER:
            if (!e->value.self) {
                MNCL_DATA *insts;
                mncl_draw_rect(132, 152, 503, 176, 128, 128, 128);
                insts = mncl_data_resource("instructions");
                if (insts && insts->tag == MNCL_DATA_ARRAY) {
                    int i, y = 156;
                    for (i = 0; i < insts->value.array.size; ++i) {
                        MNCL_DATA *val = insts->value.array.data[i];
                        mncl_draw_string(monospace, 136, y, (val && val->tag == MNCL_DATA_STRING) ? val->value.string : "(invalid datum)");
                        y += 21;
                    }
                }
            }
            break;
        default:
            break;
        }
    }

    mncl_stop_music();
    mncl_uninit();
    return 0;
}

File: pymonocle/earthball.py
import random

from pymonocle import Monocle, keycodes


mncl = Monocle()


class Globe(object):
    def __init__(self, x_max, y_max, sprite):
        self.sprite = sprite
        width = height = 64
        nframes = 30
        self.x_max = x_max - width
        self.y_max = y_max - height

        self.obj = mncl.create_object()
        self.obj.set_sprite(sprite)
        self.obj.x = random.randint(0, self.x_max)
        self.obj.y = random.randint(0, self.y_max)
        self.obj.f = random.choice(range(nframes))
        self.obj.dx = random.randint(1, 5) * random.choice([-1, 1])
        self.obj.dy = random.randint(1, 5) * random.choice([-1, 1])
        self.obj.df = 1

    def update(self):
        # Monocle actually does the movement. We just set sensible boundary
        # conditions here.
        newx = self.obj.x + self.obj.dx
        newy = self.obj.y + self.obj.dy

        if newx < 0:
            self.obj.x = self.obj.dx
            self.obj.dx = -self.obj.dx
        elif newx > self.x_max:
            self.obj.x = self.x_max + self.obj.dx
            self.obj.dx = -self.obj.dx

        if newy < 0:
            self.obj.y = self.obj.dy
            self.obj.dy = -self.obj.dy
        elif newy > self.y_max:
            self.obj.y = self.y_max + self.obj.dy
            self.obj.dy = -self.obj.dy


class EarthBall(object):

    def instructions(self):
        insts = mncl.data_resource("instructions")
        if insts.value:
            print "Instructions:"
            for val in insts.value:
                if isinstance(val.value, basestring):
                    print "    %s" % val.value
                else:
                    print "    (invalid datum)"
        else:
            print "No instructions available"

    def update(self):
        for globe in self.globes:
            globe.update()

    def render(self):
        mncl.draw_rect(256, 112, 256, 256, 128, 128, 128)

    def run(self):

        mncl.init()
        mncl.config_video("Earthball Demo", 768, 480, 0, 0)
        mncl.add_resource_zipfile("bin/earthball-res.zip")

        mncl.load_resmap("earthball.json")
        self.earth = mncl.sprite_resource("earth")
        self.sfx = mncl.sfx_resource("sfx")
        mncl.play_music_resource("bgm", 2000)

        self.globes = [Globe(768, 480, self.earth) for _ in xrange(16)]

        done = False
        countdown = 0
        music_on = True
        bg = 0
        music_volume = 128

        self.instructions()
        mncl.hide_mouse_in_fullscreen(True)

        while not done:
            e = mncl.pop_global_event()
            if e.type == mncl.MNCL_EVENT_QUIT:
                done = True

            elif e.type == mncl.MNCL_EVENT_KEYDOWN:
                if e.value.key == keycodes.SDLK_ESCAPE:
                    mncl.fade_out_music(1000)
                    countdown = 100

                elif e.value.key == keycodes.SDLK_b:
                    bg = (bg + 1) % 4
                    mncl.set_clear_color(
                        128 if bg == 1 else 0,
                        128 if bg == 2 else 0,
                        128 if bg == 3 else 0)

                elif e.value.key == keycodes.SDLK_p:
                    music_on = not music_on
                    if music_on:
                        mncl.resume_music()
                    else:
                        mncl.pause_music()

                elif e.value.key == keycodes.SDLK_t:
                    print "Fullscreen does strange things, ignored."
                    # print "Toggled fullscreen to: %s" % (
                    #     mncl.toggle_fullscreen(),)
                    # print "Fullscreen reported as: %s" % (
                    #     mncl.is_fullscreen(),)

                elif e.value.key == keycodes.SDLK_UP:
                    music_volume = min(music_volume + 16, 128)
                    mncl.music_volume(music_volume)
                    print "Music volume now: %d" % (music_volume,)

                elif e.value.key == keycodes.SDLK_DOWN:
                    music_volume = max(music_volume - 16, 0)
                    mncl.music_volume(music_volume)
                    print "Music volume now: %d" % (music_volume,)

                elif e.value.key == keycodes.SDLK_SPACE:
                    mncl.play_sfx(self.sfx, 128)

                elif e.value.key == keycodes.SDLK_LSHIFT:
                    mncl.play_sfx(self.sfx, 64)

            elif e.type == mncl.MNCL_EVENT_UPDATE:
                self.update()
                if countdown > 0:
                    countdown -= 1
                    if countdown == 0:
                        done = True

            elif e.type == mncl.MNCL_EVENT_RENDER:
                if e.value.self == mncl.NULL:
                    self.render()

        mncl.stop_music()
        mncl.uninit()


if __name__ == '__main__':
    EarthBall().run()
